optuna/samplers/_grid.py

```diff
@@ -108,18 +111,33 @@ def sample_relative(self, study, trial, search_space):
         # object is hard to get at the beginning of trial, while we need the access to the object
         # to validate the sampled value.
 
-        unvisited_grids = self._get_unvisited_grid_ids(study)
+        target_grids = self._get_unvisited_grid_ids(study)
 
-        if len(unvisited_grids) == 0:
-            raise ValueError(
-                "All grids have been evaluated. If you want to avoid this error, "
-                "please make sure that unnecessary trials do not run during "
-                "optimization by properly setting `n_trials` in `study.optimize`."
+        if len(target_grids) == 0:
+            # This case may occur with distributed optimization or trial queue. If there is no
+            # target grid, `GridSampler` evaluates a visited, duplicated point with the current
+            # trial. After that, the optimization stops.
+
+            _logger.warning(
+                "`GridSampler` is evaluating a duplicated point because all grids  have been "
+                "evaluated. This may happen due to a timing issue during distributed optimization "
+                "or an unnecessary number of `n_trials`."
```

Comment:
> ... an unnecessary number of `n_trials`

Isn't this what this PR addresses, or does this refer to the case of rerunning the optimization after a study has finished (in which case a single trial per worker will start)?